src/containers/Application/ApplicationPage.tsx

```diff
@@ -2,32 +2,96 @@ import React, { useEffect, useState } from 'react'
+import { useFormElementUpdateTracker } from '../../contexts/FormElementUpdateTrackerState'
 
 interface ApplicationProps {
   structure: FullStructure
   responses?: ResponsesByCode
 }
+// TODO move to shared types if used elsewhere
+type SectionAndPage = { sectionCode: string; pageName: string } | null
+
+interface MethodToCallOnRevalidation {
+  (firstInvalidPage: SectionAndPage): void
+}
+
+const getFirstInvalidPage: (
+  fullStructure: FullStructure | undefined
+) => { sectionCode: string; pageName: string } | null = (fullStructure) => {
+  // return { sectionCode: 'S1', pageName: 'Page 2' }
+  return null
+}
+
+interface RevalidationState {
+  methodToCallOnRevalidation: MethodToCallOnRevalidation | null
+  shouldProcessValidation: boolean
+  lastRevalidationRequest: number
+}
 
 const ApplicationPage: React.FC<ApplicationProps> = ({ structure }) => {
-  const [strictSectionPage, setStrictSectionPage] = useState({ section: null, page: null })
-  const { error, isLoading, fullStructure, responsesByCode } = useGetFullApplicationStructure({
-    structure,
-  })
   const {
     userState: { currentUser },
   } = useUserState()
   const { push, query } = useRouter()
 
+  const {
+    state: { isLastElementUpdateProcessed, elementUpdatedTimestamp },
+  } = useFormElementUpdateTracker()
+
+  const [strictSectionPage, setStrictSectionPage] = useState<SectionAndPage>(null)
+  const [revalidationState, setRevalidationState] = useState<RevalidationState>({
+    methodToCallOnRevalidation: null,
+    shouldProcessValidation: false,
+    lastRevalidationRequest: Date.now(),
+  })
+
+  const shouldRevalidate = isLastElementUpdateProcessed && revalidationState.shouldProcessValidation
+  const revalidateAfterTimestamp = shouldRevalidate ? elementUpdatedTimestamp : 0
+
+  const { error, fullStructure } = useGetFullApplicationStructure({
+    structure,
+    shouldRevalidate,
+    revalidateAfterTimestamp,
+  })
+
   const currentSection = query.sectionCode
-  const currentPage = `Page ${query.page}`
+  const currentPage = `Page${query.page}`
+
+  /* Method to pass to progress bar, next button and submit button  to cause revalidation before aciton can be proceeded
+     Should always be called on submit, but only be called on next or progress bar navigation when isLinear */
+  // TODO may rename if we want to display loading modal ?
+  const requestRevalidation = (methodToCall: MethodToCallOnRevalidation) => {
+    setRevalidationState({
+      methodToCallOnRevalidation: methodToCall,
+      shouldProcessValidation: true,
+      lastRevalidationRequest: Date.now(),
+    })
+    // TODO show loading modal ?
+  }
+
+  // Revalidation Effect
+  useEffect(() => {
+    if (
+      fullStructure &&
+      revalidationState.methodToCallOnRevalidation &&
+      fullStructure.lastValidationTimestamp > revalidationState.lastRevalidationRequest
+    ) {
+      const lastValidPage = getFirstInvalidPage(fullStructure)
```

Comment:
```suggestion
      const firstInvalidPage = getFirstInvalidPage(fullStructure)
```

to keep consistency with method name

Reply:
for sure, sorry, nice catch